torch/_functorch/partitioners.py

```diff
@@ -888,6 +906,89 @@ def get_node_weight(node) -> int:
+    # there's an unfusible op along the chain somewhere
+
+    # It could probably be improved by properly analyzing what's going on in the
+    # backwards pass instead of only relying on whether it's unfusible in the
+    # forwards.
+
+    def find_first_unfusible(start_nodes: List[fx.Node], max_range: int) -> int:
+        """
+        Finds the first unfusible node in the chain of nodes starting from
+        `start_nodes` and returns its position.
+        """
+        sorted_nodes = []
+        for n in start_nodes:
+            heapq.heappush(sorted_nodes, (n.fw_order, n, True))
+
+        while len(sorted_nodes) > 0:
+            _, node, node_is_fusible = heapq.heappop(sorted_nodes)
+            if not node_is_fusible:
+                return node.fw_order
+            if node.fw_order > max_range:
+                continue
+            for user in node.users:
+                if user in required_fw_nodes:
+                    heapq.heappush(sorted_nodes, (user.fw_order, user, is_fusible(node, user)))
+        return max_range
+
+    if BAN_IF_USED_FAR_APART:
+        for used_node in required_fw_nodes:
+            orders = [user.fw_order for user in used_node.users if user in required_fw_nodes]
+            fw_users = [user for user in used_node.users if user in required_fw_nodes]
+            if len(orders) > 0:
+                first_unfusible_use = find_first_unfusible(fw_users, max(orders))
+                for user in tuple(used_node.users):
+                    if user in required_fw_nodes and user.fw_order > first_unfusible_use and is_fusible(used_node, user):
+                        if user in banned_nodes:
+                            continue
+                        log.info(
+                            "used above/below fusible %s:(%s) -> %s -> %s:(%s)",
+                            used_node, used_node.fw_order, first_unfusible_use, user, user.fw_order
+                        )
+                        ban_recomputation_if_allowed(user)
+
+
+    # This heuristic is fairly straightforward. The idea is that although it is
+    # cheap to recompute bandwidth-bound ops, we don't want to end up in a situation
+    # where we have a long chain of pointwise ops from the beginning to the end
+    # of the model (like say, residual connections)
+
+    # It's possible this heuristic is obsoleted by the below one - checking that now
+
+    if BAN_IF_LONG_FUSIBLE_CHAINS:
+        visited = set()
+        for start_node in joint_graph.nodes:
+            if start_node not in required_fw_nodes:
+                continue
+            fusible = [(start_node.fw_order, start_node)]
+            start_pos = start_node.fw_order
+            start_order = start_node.fw_order
+            while len(fusible) > 0:
+                _, cur = heapq.heappop(fusible)
+                if cur in visited:
+                    continue
+                visited.add(cur)
+                if cur.fw_order > start_order + 100 and len(fusible) == 0:
```

Comment:
100 here does not necessarily means the chains contains 100 nodes right? Most likely the length of the chain is much less than 100.

Reply:
I think it's possible to accurately maintain the chain length similar to how `get_depth` works. Not sure if that's necessary though

Reply:
> 100 here does not necessarily means the chains contains 100 nodes right? 

Yeah, I don't think this heuristic is that important actually.